[Core] Allow locking/unlocking of `MutexLock`

AThousandShips opened the pull request:
Enables more safe use of RAII with mutexes, fixes some edge cases
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips committed:
[Core] Allow locking/unlocking of `MutexLock`

AThousandShips commented on core/io/resource_loader.cpp:
```diff
@@ -633,9 +633,9 @@ Ref<Resource> ResourceLoader::load_threaded_get(const String &p_path, Error *r_e
+				thread_load_lock.unlock();
 				OS::get_singleton()->delay_usec(1000);
-				new (&thread_load_lock) MutexLock(thread_load_mutex);
+				thread_load_lock.lock();
```
This part was the primary incentive now, where the explicit usage was presented

AThousandShips commented on core/io/resource_loader.cpp:
```diff
@@ -691,7 +691,7 @@ Ref<Resource> ResourceLoader::_load_complete_inner(LoadToken &p_load_token, Erro
 			if (loader_is_wtp) {
 				// Loading thread is in the worker pool.
-				thread_load_mutex.unlock();
+				p_thread_load_lock.unlock();
```
This helps with safety as this would have the lock be locked *and* the mutex itself unlocked, this ensures the mutex isn't operated on while the lock is existing

kus04e4ek replied:
Do changes here account for this: https://github.com/godotengine/godot/pull/93739#discussion_r1659827245?

AThousandShips replied:
The unlock is equivalent so there's no difference, and see below

kus04e4ek commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,44 @@ template <int Tag>
+	}
+
 	_ALWAYS_INLINE_ ~MutexLock() {
 		SafeBinaryMutex<Tag>::count--;
```
This implies that `MutexLock` must be locked when it hits the destructor, is this intentional? If yes, I think there should be a note somewhere about it and `DEV_ASSERT(SafeBinaryMutex<Tag>::count)` or something like this in the destructor

AThousandShips replied:
There is, in the Mutex destructor, the lock isn't responsible for this

AThousandShips replied:
But regardless this PR didn't introduce this particular code so I think if there's some issue with the structure here it'd be best to open an issue and make a dedicated fix/improvement

kus04e4ek replied:
> There is, in the Mutex destructor, the lock isn't responsible for this, it's only when the Mutex is destroyed that this should be checked

I meant `DEV_ASSERT(SafeBinaryMutex<Tag>::count)` before `SafeBinaryMutex<Tag>::count--`, so there wouldn't be negative numbers

AThousandShips replied:
That's a good point, will look

AThousandShips replied:
Added that and an additional improvement as a separate commit

AThousandShips commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,44 @@ template <int Tag>
+				return false;
+			}
+		}
+	}
```
This is taken directly from the mutex class, it should be equivalent but it's not used anywhere so hard to confirm

I added the try methods for parity with the mutexes, can remove if desired

AThousandShips commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,50 @@ template <int Tag>
+	mutable THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> _lock;
 
 public:
+	// Defer locking to ensure locking is done correctly.
```
Can split these changes off into a separate PR if desired, or drop them, just a possible improvement

AThousandShips commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,50 @@ template <int Tag>
+	_ALWAYS_INLINE_ void unlock() const {
+		DEV_ASSERT(SafeBinaryMutex<Tag>::count);
+		if (--SafeBinaryMutex<Tag>::count == 0) {
+			_lock.unlock();
```
Might need to add a `owns_mutex` here as well, will see

AThousandShips commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,51 @@ template <int Tag>
+	_ALWAYS_INLINE_ void unlock() const {
+		DEV_ASSERT(SafeBinaryMutex<Tag>::count);
+		if (--SafeBinaryMutex<Tag>::count == 0) {
+			DEV_ASSERT(_lock.owns_lock());
```
Just a check to make debugging easier, this will fail with unclear errors in some contexts

RandomShaper commented:
Overall, I like the new changes. I'd just would like to suggest making the APIs for locking-unlocking a lock more distinct so while skimming through code one knows at a glance if a certain code line is dealing with a mutex or a lock. For instance:
- `temp_unlock()`, `temp_lock()`.
- `delock()`, `relock()`.

AThousandShips commented:
Will look at some renames, thank you!

AThousandShips commented:
Changed to `temp_*` to see, will see if I remove the `try_lock` part entirely as it's currently unused and can be implemented again later

RandomShaper commented on core/os/safe_binary_mutex.h:
```diff
@@ -93,16 +93,39 @@ template <int Tag>
 	friend class ConditionVariable;
 
-	THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> lock;
+	mutable THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> _lock;
```
Despite a few Godot classes are using leading underscore for private "fields," I think the most established rule is to use that only for function names.

AThousandShips replied:
Was only named such because of the name clash with `lock()`, now it's unnecessary so will rename it

RandomShaper commented:
Looks good to me. My only remaining suggestion would be to rename `temp_lock()` to `temp_relock()` so it's clearer that a `temp_unlock()` must have happened earlier somewhere.

akien-mga commented:
Thanks!

AThousandShips commented:
Thank *you*!

AThousandShips commented:
Realized I removed some necessary checks when I rewrote the code to adjust to the new Mutex changes, will re-add them to ensure behavior when having unlocked the mutex and it is then destroyed, will push momentarily
